pkg/scaffold/olm-catalog/csv.go

```diff
@@ -0,0 +1,328 @@
+		errsb.WriteString("\n\tkind")
+	}
+	if csv.ObjectMeta.Name == "" {
+		errsb.WriteString("\n\tmetadata.name")
+	}
+	// Spec fields
+	if csv.Spec.Version.String() == "" {
+		errsb.WriteString("\n\tspec.version")
+	}
+	if csv.Spec.DisplayName == "" {
+		errsb.WriteString("\n\tspec.displayName")
+	}
+	if csv.Spec.Description == "" {
+		errsb.WriteString("\n\tspec.description")
+	}
+	if len(csv.Spec.Keywords) == 0 {
+		errsb.WriteString("\n\tspec.keywords")
+	}
+	if len(csv.Spec.Maintainers) == 0 {
+		errsb.WriteString("\n\tspec.maintainers")
+	}
+	if csv.Spec.Provider == (olmApi.AppLink{}) {
+		errsb.WriteString("\n\tspec.provider")
+	}
+	if len(csv.Spec.Labels) == 0 {
+		errsb.WriteString("\n\tspec.labels")
+	}
+
+	if len(errsb.String()) == 0 {
+		return nil
+	}
+	return errors.New(errsb.String())
+}
+
+// updateCSVVersions updates csv's version and data involving the version,
+// ex. ObjectMeta.Name, and place the old version in the `replaces` object,
+// if there is an old version to replace.
+func (s *CSV) updateCSVVersions(csv *olmApi.ClusterServiceVersion) error {
+
+	// Old csv version to replace, and updated csv version.
+	oldVer, newVer := csv.Spec.Version.String(), s.CSVVersion
+	if oldVer == newVer {
+		return nil
+	}
+
+	// We do not want to update versions in most fields, as these versions are
+	// independent of global csv version and will be updated elsewhere.
+	fieldsToUpdate := []interface{}{
+		&csv.ObjectMeta,
+		&csv.Spec.Labels,
+		&csv.Spec.Selector,
+	}
+	for _, v := range fieldsToUpdate {
+		err := replaceAllBytes(v, []byte(oldVer), []byte(newVer))
+		if err != nil {
+			return err
+		}
+	}
+
+	// Now replace all references to the old operator name.
+	lowerProjName := strings.ToLower(s.ProjectName)
+	oldCSVName := getCSVName(lowerProjName, oldVer)
+	newCSVName := getCSVName(lowerProjName, newVer)
+	err := replaceAllBytes(csv, []byte(oldCSVName), []byte(newCSVName))
+	if err != nil {
+		return err
+	}
+
+	csv.Spec.Version = *semver.New(newVer)
+	csv.Spec.Replaces = oldCSVName
+	return nil
+}
+
+func replaceAllBytes(v interface{}, old, new []byte) error {
+	b, err := json.Marshal(v)
+	if err != nil {
+		return err
+	}
+	b = bytes.Replace(b, old, new, -1)
+	if err = json.Unmarshal(b, v); err != nil {
+		return err
+	}
+	return nil
+}
+
+// updateCSVFromManifestFiles gathers relevant data from generated and user-defined manifests
+// and updates csv.
+func (s *CSV) updateCSVFromManifestFiles(cfg *CSVConfig, csv *olmApi.ClusterServiceVersion) error {
+	store := NewUpdaterStore()
+	for _, f := range append(cfg.CRDCRPaths, cfg.OperatorPath, cfg.RolePath) {
+		yamlData, err := ioutil.ReadFile(f)
+		if err != nil {
+			return err
+		}
+
+		scanner := yamlutil.NewYAMLScanner(yamlData)
+		for scanner.Scan() {
+			yamlSpec := scanner.Bytes()
+
+			k, err := getKindfromYAML(yamlSpec)
+			if err != nil {
+				return err
+			}
+
+			updateFunc := getUpdaterFunc(store, k)
+			if updateFunc != nil {
+				updateFunc(yamlSpec)
```

Comment:
Check return error?

Reply:
Also I'm wondering if we can't combine `getUpdaterFunc()` and `updateFunc()` into a single method for `updaterStore`.
E.g:
```Go
err := store.AddUpdater(yamlSpec)
```
which parses the yaml for the kind and calls the correct updateFunc e.g `AddRole` for the store. That's essentially what we're doing now but it took me a while to understand that.

Reply:
I've combined both calls and removed `CSVUpdaterSet` in favor of using only `CSVUpdater` and `updateStore`, in which `CSVUpdaterSet.Apply()` -> `updateStore.Apply()`.